### PR TITLE
[WIP] - ecs support

### PR DESCRIPTION
**Pull request by garethmdavies**
Hi @brabster,

Whilst I was at the AWS builders container event in Manchester yesterday (very good event) I decided to add the ecs-support to crucible. We are looking at moving the web/mid tier platform to containers possibly fargate when it becomes available. I'm looking at comparing crucible to terraform for the cloudformation templates for this bit of work.

Once I did the code changes I realised @keerts was already adding this support but had gaps in the container definition. I thought I would create a pull request anyway in case what I've done saves @keerts some time. I'm missing some tests but have added, cluster, service and task definitions specs. Probably could do with some sensible defaults for some of the specs as well. I also found it easier to read by splitting the code into different namespaces especially when there are multiple resources using properties with the same name but of different type e.g. cpu

cheers

### CODE REVIEW

**Comment by brabster** [comment]
Nice @taffowl, @keerts you OK if I merge this?

**Comment by keerts** [comment]
Splitting up into different namespaces is a good alternative to what I did to resolve the problem of duplicate property names:

(s/def :crucible.aws.ecs.task-definition.key-value-pair/name (spec-or-ref string?))
(s/def :crucible.aws.ecs.task-definition.key-value-pair/value (spec-or-ref string?))
(s/def ::key-value-pair (s/keys :req [:crucible.aws.ecs.task-definition.key-value-pair/name
                                      :crucible.aws.ecs.task-definition.key-value-pair/value]))

This is inherently hard when writing the cloudformation template.

When we choose to go this route, we have to split existing namespaces, like lambda, as well to make it consistent.

File: src/crucible/aws/ecs.clj [hunk 1]
```diff
@@ -0,0 +1,19 @@
+(ns crucible.aws.ecs
+  "Resources in AWS::ECS::*"
+  (:require [clojure.spec.alpha :as s]
+            [crucible.resources :refer [spec-or-ref defresource]]
+            [crucible.aws.ecs.task-definition :as td]
+            [crucible.aws.ecs.service :as svc]))
+
+(defn ecs [suffix] (str "AWS::ECS::" suffix))
+
+(s/def ::cluster-name (spec-or-ref string?))
```
Comment (keerts) [inline src/crucible/aws/ecs.clj]: Although it is only one, I would move this into it's own namespace as well for consistency.

File: src/crucible/aws/ecs/container_definition.clj [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+(s/def ::mount-points (s/keys :req [::container-path
+                                    ::source-volume]
+                              :opt [::read-only]))
+
+(s/def ::container-port (spec-or-ref integer?))
+(s/def ::host-port (spec-or-ref integer?))
+(s/def ::protocol (spec-or-ref string?))
+
+(s/def ::port-mappings (s/keys :req [::container-port]
```
Comment (keerts) [inline src/crucible/aws/ecs/container_definition.clj]: I believe this should be:

(s/def ::port-mapping (s/keys :req [::container-port]
                              :opt [::host-port
                                    ::protocol]))
(s/def ::port-mappings (s/* ::port-mapping))

File: src/crucible/aws/ecs.clj [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+(s/def ::cluster (s/keys :req []
+                     :opt [::cluster-name]))
+
+(defresource cluster (ecs "Cluster") ::cluster)
+
+(defresource service (ecs "Service") ::svc/service)
+
+(defresource task-definition (ecs "TaskDefinition") ::td/task-definition)
```
Comment (keerts) [inline src/crucible/aws/ecs.clj]: Newline missing at end of file (all files in the pr have this)

File: src/crucible/aws/ecs/service.clj [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                                            ::container-port
+                                            ::load-balancer-name
+                                            ::target-group-arn]))
+
+(s/def ::load-balancers (s/coll-of ::single-load-balancer :kind vector?))
+
+
+(s/def ::subnets (s/coll-of (spec-or-ref string?) :kind vector?))
+
+(s/def ::assign-public-ip (spec-or-ref string?))
```
Comment (keerts) [inline src/crucible/aws/ecs/service.clj]: Could be more specific:

(s/def ::assign-public-ip #{"ENABLED" "DISABLED"})

(But I must admit, I did this here and there, but could probably be more consistent ;-) )

File: src/crucible/aws/ecs/task_definition.clj [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+(s/def ::cpu (spec-or-ref string?))
+
+(s/def ::execution-role-arn (spec-or-ref string?))
+
+(s/def ::family (spec-or-ref string?))
+
+(s/def ::memory (spec-or-ref string?))
+
+(s/def ::network-mode (spec-or-ref string?))
```
Comment (keerts) [inline src/crucible/aws/ecs/task_definition.clj]: Could be more specific:

(s/def ::network-mode #{"bridge" "host" "awsvpc" "none"})

File: src/crucible/aws/ecs/task_definition.clj [hunk 6]
```diff
@@ -0,0 +1,40 @@
+(ns crucible.aws.ecs.task-definition
+  "Resources in AWS::ECS::TaskDefinition"
+  (:require [clojure.spec.alpha :as s]
+            [crucible.resources :refer [spec-or-ref defresource]]
+            [crucible.aws.ecs.container-definition :as cd]))
+
+(s/def ::container-definitions (s/coll-of ::cd/container-definition :kind vector?))
+
+(s/def ::cpu (spec-or-ref string?))
```
Comment (keerts) [inline src/crucible/aws/ecs/task_definition.clj:9]: Could be more specific:

(s/def ::cpu #{256 512 1024 2048 4096})

**Review by keerts — CHANGES_REQUESTED** [review]
Nice work! It is more complete than what I had.

File: src/crucible/aws/ecs/task_definition.clj (hunk 6 above)
Comment (garethmdavies) [reply src/crucible/aws/ecs/task_definition.clj:9]: For a Fargate launch type you are correct and are limited to the above values as strings however for a normal ec2 user the value is optional and can have any value. I'm not sure how to make the spec conditional on the launch type though.

Comment (keerts) [reply src/crucible/aws/ecs/task_definition.clj:9]: Ah right, good catch. I'm on FARGATE obviously ;-)
There are more specs that can be made conditionally based on other values, we'll have to learn spec a bit better. You are obviously already doing better than I am.

**Review by keerts — APPROVED** [review]

**Comment by brabster** [comment]
Available in 0.30.0

**Pull request merged**